### PR TITLE
Extracted a writer class from runner

### PR DESCRIPTION
**Pull request by SadCheeto**
## Describe your changes
Extracted a writer class from runner to ensure that runner is only in charge of actually running the aggregations.
Also moved the different save functions out into their own functions. 

NOTE: run and _write_to_csv could probably be merged - I would love to hear your thoughts
## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [x] Have I requested and notified the boys for review of the pull-request?

### CODE REVIEW

File: byoqm/writer/writer.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        run ensures that the output path exists and thereafter starts saving the
+        respective files
+        """
+        if results == None:
+            logging.error("Results is none. Skipping writing...")
+            return
+        self._gen_output_paths_if_not_exists()
+        self._write_to_csv(results)
+
+    def _write_to_csv(self, results: Dict):
```
Comment (toffernator) [inline byoqm/writer/writer.py:36]: At this point we should name it as if it is a public method i.e. drop `_` as a prefix.

File: byoqm/writer/writer.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        # explaining formattng
+        # Format: YYYY-MM-DD_HH-MM-SS
+        logging.info("Writing to csv")
+
+        self._save_meta_data()
+        self._save_frequencies(results)
+        self._save_violations(results)
+        logging.info("Finished writing to csv")
+
+    def _save_meta_data(self):
```
Comment (toffernator) [inline byoqm/writer/writer.py:58]: Same comment on naming

File: byoqm/writer/writer.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        quality model used and the src path
+        """
+        file_location = self._output_dir / Path("metadata") / self._file_name
+        with open(file_location, "w") as metadata_file:
+            writer = csv.writer(metadata_file)
+            writer.writerow([f"qualitymodel", "src_root"])
+            writer.writerow([self._model_name, self._shortened_path.__str__()])
+        logging.info("Finished writing metadata to csv")
+
+    def _save_frequencies(self, results):
```
Comment (toffernator) [inline byoqm/writer/writer.py:70]: Same comment on naming

File: byoqm/writer/writer.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            writer = csv.writer(results_file)
+            writer.writerow(["metric", "value"])
+            for description, value in results.items():
+                frequency = value
+                if type(value) is Result:
+                    frequency = value.get_frequency()
+                writer.writerow([description, frequency])
+        logging.info("Finished writing frequencies to csv")
+
+    def _save_violations(self, results: Dict):
```
Comment (toffernator) [inline byoqm/writer/writer.py:86]: Same comment on naming

**Review by toffernator — COMMENTED** [review]
I don't think the interface makes much sense with `run` being the only "public" (denoted by its naming) function and everything else being private. Also this would lead to code like `writer.run()` which, doesn't describe what it is doing very well.

File: byoqm/main.py [hunk 5]
```diff
@@ -77,16 +78,17 @@ def load(
...
-        save_file,
         language.lower(),
     )
-    runner.run()
-    logging.info("Finished running")
+    results = runner.run()
+
+    writer: Writer = Writer()
+    if save_file:
+        writer.run(results, output, quality_model, src_root)
```
Comment (toffernator) [inline byoqm/main.py:91]: Here I think syntax like

```python
writer.write_metadata(...)
writer.write_frequencies(...)
writer.write_violations(...)
```

or 

```python
writer.write_csv(...)
```

Comment (SadCheeto) [reply byoqm/main.py]: 2nd option

File: byoqm/writer/test_writer.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+class TestRunner(unittest.TestCase):
+    def setUp(self):
+        self._writer = Writer()
+        logging.disable()
+
+    def tearDown(self):
+        # shutil over os.rmdir, because it allows you to remove non/empty directories
+        shutil.rmtree(_TEST_FOLDER, ignore_errors=True)
+
+    def test_run_produces_non_empty_file(self):
```
Comment (toffernator) [inline byoqm/writer/test_writer.py:29]:
```suggestion
    def test_write_to_csv_produces_non_empty_file(self):
```

**Review by toffernator — COMMENTED** [review]
Git is trying to automatically delete empty files, be careful og `git add .`. Also added a single suggestion

**Review by toffernator — APPROVED** [review]

**Pull request merged**